web/app/solar.js

```diff
@@ -56,72 +71,116 @@ exports.draw = function(canvasSelector, now, grib1, grib2, solarColor, projectio
+    var S = projection.invert([realW / 2, realH])
+    var N = projection.invert([realW / 2, 0])
+    N[1] = 80;
+
+    var minLat = Math.ceil(SE[1]);
+    var maxLat = Math.floor(N[1]);
+    var minLon = Math.ceil(NW[0]);
+    var maxLon = Math.floor(NE[0]);
+    h = maxLat - minLat;
+    w = maxLon - minLon;
+
+    var dt = new Date().getTime();
+    console.log('Draw solar start:' + new Date())
+
+    // Draw initial image (1px 1deg) from grib
+    var imgGrib = ctx.createImageData(w, h);
+    d3.range(minLat, maxLat).forEach(function (y) {
+        d3.range(minLon, maxLon).forEach(function (x) {
+
+            var i = parseInt(x - lo1) / dx;
+            var j = parseInt(la1 - y) / dy;
+            var n = i + Nx * j;
+            var val = d3.interpolate(grib1.data[n], grib2.data[n])(k);
+
+            var pix_x = x - minLon;
+            var pix_y = h - (y - minLat);
+
+            if (levels)
+                imgGrib.data[[((pix_y * (imgGrib.width * 4)) + (pix_x * 4)) + 3]] = parseInt(val / solarScale * 255);
+            else
+                imgGrib.data[[((pix_y * (imgGrib.width * 4)) + (pix_x * 4)) + 3]] = parseInt((0.85 - val / solarScale) * 255);
         });
+    });
+    solarCanvas.attr('height', h);
+    solarCanvas.attr('width', w);
+
+    ctx.clearRect(0, 0, h, w);
+    ctx.putImageData(imgGrib, 0, 0);
+
+    console.log('Extract grib:' + (new Date().getTime() - dt));
+    dt = new Date().getTime();
+
+    // Reproject this image on our projection
+    var sourceData = ctx.getImageData(0, 0, w, h).data,
+        target = ctx.createImageData(realW, realH),
+        targetData = target.data;
 
-        if (x + batchsize <= xrange[xrange.length - 1]) {
-            setTimeout(function() { return batchDrawColumns(x + batchsize, batchsize, callback); }, 25);
-        } else {
-            console.log('Done drawing solar');
-            callback(null);
+    for (var y = 0, i = -1; y < realH; ++y) {
+        for (var x = 0; x < realW; ++x) {
+            var p = projection.invert([x, y]), lon = p[0], lat = p[1];
+            if (lon > maxLon || lon < minLon || lat > maxLat || lat < minLat) { i += 4; continue; }
+            var q = ((maxLat - lat) / (maxLat - minLat) * h | 0) * w + ((lon - minLon) / (maxLon - minLon) * w | 0) << 2;
+            i += 3;
+            q += 2;
+            targetData[++i] = sourceData[++q];
         }
     }
 
-    batchDrawColumns(0, 200, function() {
-        ctx.clearRect(0, 0, w, h);
-        ctx.putImageData(img, 0, 0);
-        callback(null);
-    });
+    ctx.clearRect(0, 0, w, h);
+    solarCanvas.attr('height', realH);
+    solarCanvas.attr('width', realW);
+    ctx.putImageData(target, 1, 1);
+    console.log('Reroject time:' + (new Date().getTime() - dt));
+
+    // Apply a gaussian blur on grid cells
+    if (gaussianBlur) {
+        dt = new Date().getTime();
+
+        stackBlurCanvasRGBA(solarCanvas._groups[0][0].id, 0, 0, realW, realH, BLUR_RADIUS);
```

Comment:
`solarCanvas._groups[0][0].id` is undefined. We need to give the solarcanvas an `id` attr if you need to make it that way. Is there another way to pass the reference? Else we change the `canvasSelector` in this class by a `canvasId`